src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectTypes/Database/DatabaseHandler.cs

```diff
@@ -210,39 +219,141 @@ public override Task Save(DatabaseViewContext context, DatabaseInfo obj)
+                            smoDatabase.Alter(SqlServer.Management.Smo.TerminationClause.RollbackTransactionsImmediately);
+                        }
+                        dataContainer.Server!.DetachDatabase(smoDatabase.Name, updateStatistics);
+                    }
+                    catch (SmoException ex)
+                    {
+                        // Revert to previous level of user access if we changed it before encountering the exception
+                        if (originalAccess != smoDatabase.DatabaseOptions.UserAccess)
+                        {
+                            smoDatabase.DatabaseOptions.UserAccess = originalAccess;
+                            smoDatabase.Alter(TerminationClause.RollbackTransactionsImmediately);
+                        }
+                        if (!(ex.InnerException is MissingObjectException) || (ex.InnerException is MissingObjectException && throwIfNotExist))
+                        {
+                            throw;
+                        }
+                    }
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Provided URN '{objectUrn}' did not correspond to an existing database.");
+                }
+            }
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Used to produce a TSQL script for detaching the specified database from a server.
+        /// </summary>
+        /// <param name="connectionUri">URI of the underlying connection for the detach request</param>
+        /// <param name="objectUrn">URN of the database to detach</param>
+        /// <param name="dropConnections">Whether to drop active connections to the database before detaching it</param>
+        /// <param name="updateStatistics">Whether to update the query optimization statistics related to the database</param>
+        /// <param name="throwIfNotExist">Whether to throw an exception if the specified database doesn't exist</param>
+        /// <returns>A string representing the generated TSQL script</returns>
+        public string ScriptDetach(string connectionUri, string objectUrn, bool dropConnections, bool updateStatistics, bool throwIfNotExist)
+        {
+            var builder = new StringBuilder();
+            ConnectionInfo connectionInfo = this.GetConnectionInfo(connectionUri);
+            using (var dataContainer = CreateDatabaseDataContainer(connectionUri, objectUrn))
+            {
+                try
+                {
+                    var smoDatabase = dataContainer.SqlDialogSubject as Database;
+                    if (smoDatabase != null)
+                    {
+                        builder.AppendLine("USE [master]");
+                        builder.AppendLine("GO");
+                        if (dropConnections)
+                        {
+                            builder.AppendLine($"ALTER DATABASE [{smoDatabase.Name}] SET SINGLE_USER WITH ROLLBACK IMMEDIATE");
```

Comment:
You need to escape the database name here

Reply:
Even if it comes from the SMO Database object?

Reply:
I retrieve the Database object using the object URN, rather than using any arguments directly.

Reply:
That name isn't going to be escape...so yes. A database named `MyTest]Database` will break the query currently.